pkg/job/scraper.go

```diff
@@ -0,0 +1,240 @@
+				return ""
+			}
+			return accountID
+		})
+	})
+
+	var wg sync.WaitGroup
+	mux := &sync.Mutex{}
+	jobErrors := make([]Error, 0)
+	metricResults := make([]model.CloudwatchMetricResult, 0)
+	resourceResults := make([]model.TaggedResourceResult, 0)
+	s.logger.Debug("Starting job runs")
+
+	jobConfigVisitor(s.jobsCfg, func(job any, role model.Role, region string) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			var namespace string
+			jobAction(s.logger, job, func(job model.DiscoveryJob) {
+				namespace = job.Type
+			}, func(job model.CustomNamespaceJob) {
+				namespace = job.Namespace
+			})
+			jobLogger := s.logger.With("namespace", namespace, "region", region, "arn", role.RoleArn)
+
+			accountID := roleRegionToAccount[role][region]()
+			if accountID == "" {
+				jobError := Error{
+					AccountID: "",
+					Region:    region,
+					RoleARN:   role.RoleArn,
+					Namespace: namespace,
+					Message:   "Account for job was not found see previous errors",
+				}
+				mux.Lock()
+				jobErrors = append(jobErrors, jobError)
+				mux.Unlock()
+				return
+			}
+			jobLogger = jobLogger.With("account", accountID)
+
+			var jobToRun cloudwatchrunner.Job
+			jobAction(jobLogger, job,
+				func(job model.DiscoveryJob) {
+					jobLogger.Debug("Starting resource discovery")
+					rmRunner := s.runnerFactory.NewResourceMetadataRunner(jobLogger, region, role)
+					resources, err := rmRunner.Run(ctx, region, job)
+					if err != nil {
+						jobError := Error{
+							AccountID: accountID,
+							Region:    region,
+							RoleARN:   role.RoleArn,
+							Namespace: namespace,
+							Message:   "Failed to run resource metadata for job",
+							Err:       err,
+						}
+						mux.Lock()
+						jobErrors = append(jobErrors, jobError)
+						mux.Unlock()
+
+						return
+					}
+					if len(resources) > 0 {
+						result := model.TaggedResourceResult{
+							Context: &model.ScrapeContext{
+								Region:     region,
+								AccountID:  accountID,
+								CustomTags: job.CustomTags,
+							},
+							Data: resources,
+						}
+						mux.Lock()
+						resourceResults = append(resourceResults, result)
+						mux.Unlock()
+					} else {
+						jobLogger.Debug("No tagged resources")
+					}
+					jobLogger.Debug("Resource discovery finished", "number_of_discovered_resources", len(resources))
+
+					jobToRun = cloudwatchrunner.DiscoveryJob{Job: job, Resources: resources}
+				}, func(job model.CustomNamespaceJob) {
+					jobToRun = cloudwatchrunner.CustomNamespaceJob{Job: job}
+				},
+			)
+			if jobToRun == nil {
+				jobLogger.Debug("Ending job run early due to job error see job errors")
+				return
+			}
+			jobLogger.Debug("Starting cloudwatch metrics runner")
+			cwRunner := s.runnerFactory.NewCloudWatchRunner(jobLogger, region, role, jobToRun)
+			metricResult, err := cwRunner.Run(ctx)
+			if err != nil {
+				jobError := Error{
+					AccountID: accountID,
+					Region:    region,
```

Comment:
I noticed that a few of the params are reused in this struct throughout this function, maybe a single error instance should be built at the beginning, add the AccountID when it exists, and modify the `Message` and `Err` params as needed?

Reply:
Since the error isn't necessary for most job runs I'm fine with leaving creation for when we need it vs creating one and throwing it away most of the time.

Reply:
I ended up taking this a different direction where instead of building the Error type I build up a context object. This object is then morphed to what's needed in the return hiding a lot of the distracting mapping logic.